data/src/main/java/com/linkedin/data/codec/entitystream/AbstractJacksonDataDecoder.java

```diff
@@ -76,328 +61,147 @@
     {
-      JsonToken token;
-      while ((token = _jsonParser.nextToken()) != null)
+      if(_byteArrayFeeder.needMoreInput())
       {
-        switch (token)
-        {
-          case START_OBJECT:
-            validate(START_OBJECT);
-            // If we are already filling out a DataMap, we cannot reuse _currDataMapBuilder and thus
-            // need to create a new one.
-            if (_currDataMapBuilder.inUse())
-            {
-              _currDataMapBuilder = new DataMapBuilder();
-            }
-            _currDataMapBuilder.setInUse(true);
-            push(_currDataMapBuilder, false);
-            break;
-          case END_OBJECT:
-            validate(END_OBJECT);
-            pop();
-            break;
-          case START_ARRAY:
-            validate(START_ARRAY);
-            push(new DataList(), true);
-            break;
-          case END_ARRAY:
-            validate(END_ARRAY);
-            pop();
-            break;
-          case FIELD_NAME:
-            validate(FIELD_NAME);
-            _currField = _jsonParser.getCurrentName();
-            _expectedTokens = VALUE;
-            break;
-          case VALUE_STRING:
-            validate(SIMPLE_VALUE);
-            addValue(_jsonParser.getText());
-            break;
-          case VALUE_NUMBER_INT:
-          case VALUE_NUMBER_FLOAT:
-            validate(SIMPLE_VALUE);
-            JsonParser.NumberType numberType = _jsonParser.getNumberType();
-            switch (numberType)
-            {
-              case INT:
-                addValue(_jsonParser.getIntValue());
-                break;
-              case LONG:
-                addValue(_jsonParser.getLongValue());
-                break;
-              case FLOAT:
-                addValue(_jsonParser.getFloatValue());
-                break;
-              case DOUBLE:
-                addValue(_jsonParser.getDoubleValue());
-                break;
-              case BIG_INTEGER:
-              case BIG_DECIMAL:
-              default:
-                handleException(new Exception("Unexpected number value type " + numberType + " at " + _jsonParser.getTokenLocation()));
-                break;
-            }
-            break;
-          case VALUE_TRUE:
-            validate(SIMPLE_VALUE);
-            addValue(Boolean.TRUE);
-            break;
-          case VALUE_FALSE:
-            validate(SIMPLE_VALUE);
-            addValue(Boolean.FALSE);
-            break;
-          case VALUE_NULL:
-            validate(SIMPLE_VALUE);
-            addValue(Data.NULL);
-            break;
-          case NOT_AVAILABLE:
-            readNextChunk();
-            return;
-          default:
-            handleException(new Exception("Unexpected token " + token + " at " + _jsonParser.getTokenLocation()));
-        }
+        _byteArrayFeeder.feedInput(data, offset, end);
+      }
+      else
+      {
+        throw new IOException("Byte Array Feeder is not ok to feed more data.");
```

Comment:
I think this can be worded better. Something like
"Invalid state: Parser cannot accept more data"